feat: Add `linear_spaces`

mcrumiller opened the pull request:
Closes #20922.

### Examples

```python
import polars as pl
pl.Config.set_fmt_table_cell_list_len(5)

df = pl.DataFrame({
    "start": [1, -1],
    "end": [3, 2],
    "num_samples": [4, 5]}
)

df.with_columns(ls=pl.linear_spaces("start", "end", "num_samples"))
# shape: (2, 4)
# ┌───────┬─────┬─────────────┬────────────────────────────────┐
# │ start ┆ end ┆ num_samples ┆ ls                             │
# │ ---   ┆ --- ┆ ---         ┆ ---                            │
# │ i64   ┆ i64 ┆ i64         ┆ list[f64]                      │
# ╞═══════╪═════╪═════════════╪════════════════════════════════╡
# │ 1     ┆ 3   ┆ 4           ┆ [1.0, 1.666667, 2.333333, 3.0] │
# │ -1    ┆ 2   ┆ 5           ┆ [-1.0, -0.25, 0.5, 1.25, 2.0]  │
# └───────┴─────┴─────────────┴────────────────────────────────┘
```

Using `as_array=True` returns a `pl.Array`, but can only be used with a constant `num_samples` (integer or literal integer):

```python
df.with_columns(ls=pl.linear_spaces("start", "end", 3, as_array=True))
# shape: (2, 4)
# ┌───────┬─────┬─────────────┬──────────────────┐
# │ start ┆ end ┆ num_samples ┆ ls               │
# │ ---   ┆ --- ┆ ---         ┆ ---              │
# │ i64   ┆ i64 ┆ i64         ┆ array[f64, 3]    │
# ╞═══════╪═════╪═════════════╪══════════════════╡
# │ 1     ┆ 3   ┆ 4           ┆ [1.0, 2.0, 3.0]  │
# │ -1    ┆ 2   ┆ 5           ┆ [-1.0, 0.5, 2.0] │
# └───────┴─────┴─────────────┴──────────────────┘
```

Temporals work as well, and we can define the interval with `left`, `right`, `none`, or `both` (default):

```python
import polars as pl
from datetime import date
pl.Config.set_fmt_table_cell_list_len(4)
pl.Config.set_fmt_str_lengths(99)

df = pl.DataFrame({
    "start": [date(2025, 1, 1), date(2025, 1, 2)],
    "end": [date(2025, 2, 1), date(2025, 2, 2)],
    "num_samples": [3, 4]}
)
df.with_columns(ls=pl.linear_spaces("start", "end", 3, as_array=True, closed="left"))
# shape: (2, 4)
# ┌────────────┬────────────┬─────────────┬─────────────────────────────────────────────────────────────────┐
# │ start      ┆ end        ┆ num_samples ┆ ls                                                              │
# │ ---        ┆ ---        ┆ ---         ┆ ---                                                             │
# │ date       ┆ date       ┆ i64         ┆ array[datetime[ms], 3]                                          │
# ╞════════════╪════════════╪═════════════╪═════════════════════════════════════════════════════════════════╡
# │ 2025-01-01 ┆ 2025-02-01 ┆ 3           ┆ [2025-01-01 00:00:00, 2025-01-11 08:00:00, 2025-01-21 16:00:00] │
# │ 2025-01-02 ┆ 2025-02-02 ┆ 4           ┆ [2025-01-02 00:00:00, 2025-01-12 08:00:00, 2025-01-22 16:00:00] │
# └────────────┴────────────┴─────────────┴─────────────────────────────────────────────────────────────────┘
```

mcrumiller commented:
@coastalwhite I've added the `as_array` parameter. Let me know if you think the implementation is reasonable. To use that parameter, we have to verify that `num_samples` is a constant, which means either a python int or a `pl.lit(x, dtype=<integer dtype>)`. In order to correctly determine the schema for the `pl.Array` dtype, we have to know what the width is in advance. So we check that the `num_samples` expr is a constant literal and send along `array_width: Option<usize>` to the DSL. The implementation can then later see if this is a `Some(width)` and if so it creates the array.

coastalwhite commented on crates/polars-plan/src/dsl/functions/range.rs:
```diff
@@ -175,3 +175,64 @@ pub fn linear_space(start: Expr, end: Expr, num_samples: Expr, closed: ClosedInt
+        None
+    };
+
+    let input = vec![start, end, num_samples];
```
I think, ideally, if the array_width is `Some(..)` then we don't pass `num_samples` here. It is a redundant expression, and we don't need to look at this in the optimizer.

mcrumiller replied:
[9c490c8](https://github.com/pola-rs/polars/pull/20941/commits/9c490c81ae80667d480f98338096e3744aca367e)

coastalwhite commented on crates/polars-plan/src/dsl/functions/range.rs:
```diff
@@ -175,3 +175,64 @@ pub fn linear_space(start: Expr, end: Expr, num_samples: Expr, closed: ClosedInt
     }
 }
+
+fn match_literal_int(expr: &Expr) -> PolarsResult<Option<usize>> {
```
I think this should be an `extract_size` or `extract_usize` method on `Expr` and not here. It is more generic than that.

mcrumiller replied:
[92f03e5](https://github.com/pola-rs/polars/pull/20941/commits/92f03e5efdacfa423281b9ee6e1c6fb85474f397)

mcrumiller replied:
I'm not sure if this is the best place for it--let me know if you have a better idea.

coastalwhite commented on crates/polars-plan/src/dsl/function_expr/range/linear_space.rs:
```diff
@@ -65,7 +68,138 @@ pub(super) fn linear_space(s: &[Column], closed: ClosedInterval) -> PolarsResult
+) -> PolarsResult<Column> {
+    let start = &s[0];
+    let end = &s[1];
+    let num_samples = &s[2];
```
If the `num_spaces` is removed on creation, this will only have `s.len() == 2`.

mcrumiller replied:
[9c490c8](https://github.com/pola-rs/polars/pull/20941/commits/9c490c81ae80667d480f98338096e3744aca367e)

coastalwhite commented on crates/polars-plan/src/dsl/function_expr/range/linear_space.rs:
```diff
@@ -65,7 +68,138 @@ pub(super) fn linear_space(s: &[Column], closed: ClosedInterval) -> PolarsResult
+            let mut builder = ListPrimitiveChunkedBuilder::<Float32Type>::new(
+                name,
+                len,
+                len * CAPACITY_FACTOR,
```
I think this factor can be calculated exactly when `as_array` is `Some(..)`.

coastalwhite commented on crates/polars-plan/src/dsl/function_expr/range/linear_space.rs:
```diff
@@ -65,7 +68,138 @@ pub(super) fn linear_space(s: &[Column], closed: ClosedInterval) -> PolarsResult
+            let mut builder = ListPrimitiveChunkedBuilder::<Float64Type>::new(
+                name,
+                len,
+                len * CAPACITY_FACTOR,
```
idem.

mcrumiller replied:
[5148cf6](https://github.com/pola-rs/polars/pull/20941/commits/5148cf621ed0caba67875732eebabc0bc691cab5)

coastalwhite commented on crates/polars-plan/src/dsl/function_expr/range/linear_space.rs:
```diff
@@ -65,7 +68,138 @@ pub(super) fn linear_space(s: &[Column], closed: ClosedInterval) -> PolarsResult
+            let mut builder = ListPrimitiveChunkedBuilder::<Float64Type>::new(
+                name,
+                len,
+                len * CAPACITY_FACTOR,
```
idem.

mcrumiller replied:
[5148cf6](https://github.com/pola-rs/polars/pull/20941/commits/5148cf621ed0caba67875732eebabc0bc691cab5)

coastalwhite commented on crates/polars-plan/src/dsl/function_expr/range/utils.rs:
```diff
@@ -246,10 +247,163 @@ where
     Ok(())
 }
 
+/// Create a ranges column from the given start/end columns and a range function.
```
nit. Any reason why this is in `utils.rs` and not in `linear_space.rs`?

mcrumiller replied:
I was following the convention with the other range functions, but I agree that it should be in `linear_space`. I'ved moved it in [4ca05a9](https://github.com/pola-rs/polars/pull/20941/commits/4ca05a97fad28d6fbe68d0ce40286ecdab9d3375).

coastalwhite commented on py-polars/polars/functions/range/linear_space.py:
```diff
@@ -183,3 +189,115 @@ def linear_space(
+) -> Expr: ...
+
+
+@overload
```
I am not a python expert, but is it possible to make an overload for `as_array = True` to have `num_samples` only take the `int` type hint.

mcrumiller replied:
Hang on, getting a mypy error (not sure why it didn't trigger locally).

mcrumiller replied:
Ok, so nope. The reason is because we also allow `pl.lit(int)` with `as_array = True`.

coastalwhite replied:
Okay clear

coastalwhite commented on py-polars/polars/functions/range/linear_space.py:
```diff
@@ -183,3 +189,115 @@ def linear_space(
+    eager: bool = False,
+) -> Expr | Series:
+    """
+    Generate a sequence of evenly-spaced points for each row of the input columns.
```
I think this description could be a little bit more clear. Maybe something like (feel free to improve further):

```
Generate a sequence of evenly-spaced values for each row between the value of `start` and `end`. The number of values is determined by `num_samples`.

mcrumiller replied:
[31129c2](https://github.com/pola-rs/polars/pull/20941/commits/31129c22e41be491fa99c5f81d3e137cd1b03992)

coastalwhite commented on py-polars/polars/functions/range/linear_space.py:
```diff
@@ -183,3 +189,115 @@ def linear_space(
+    closed : {'both', 'left', 'right', 'none'}
+        Define which sides of the interval are closed (inclusive).
+    as_array
+        Return result as a fixed-length pl.Array. `num_samples` must be a constant.
```
nit. I think this `pl.Array` should be `Array` and should be code.

mcrumiller replied:
[31129c2](https://github.com/pola-rs/polars/pull/20941/commits/31129c22e41be491fa99c5f81d3e137cd1b03992)

coastalwhite reviewed (changes requested):
I really like the implementation. I left a few comments and nits.

mcrumiller commented:
@coastalwhite thanks for the thorough review. I think I've addressed all of your issues, and linked to the relevant commits above. The code coverage error is due to some network issue, not related to any changes.

coastalwhite commented on crates/polars-plan/src/dsl/expr.rs:
```diff
@@ -369,6 +369,38 @@ impl Expr {
     }
+
+    /// Extract a constant usize from an expression.
+    pub fn extract_usize(&self) -> PolarsResult<Option<usize>> {
```
Not sure why this is a `Result<Option<usize>>`, I don't think it returns `Ok(None)` anywhere

mcrumiller replied:
You're right, that was a carryover from its earlier usage as a helper function for `array_width`. I've removed the Option in [97a5a98](https://github.com/pola-rs/polars/pull/20941/commits/97a5a98d1e80e6e851b99c9a029f803017c3066e).

mcrumiller replied:
I'm wondering those `.unwrap()`s are ok or if we should return an error.

mcrumiller replied:
Mapped the error in [bfdaa7f](https://github.com/pola-rs/polars/pull/20941/commits/bfdaa7fb8ad67864948ccc020e91383a1c3f534d).

coastalwhite reviewed (changes requested):
I had one last comment.

coastalwhite commented:
@ritchie46 I think this can be merged, maybe you want a final look
